3_WordBreak.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+ Microsoft Interview Coaching Pilot Program MICP 2018
+Author: Akshama
+Language: C++
+/**********WEEK THREE***********/
+/*
+Word Break
+Given an input string and a dictionary of words, find out if the input string can be segmented into a space-separated sequence of dictionary words.
+For example, consider the following dictionary: { pear, salmon, foot, prints, footprints, leave, you, sun, girl, enjoy },
+Examples:
+Given the string �youenjoy�,
+Output: True (The string can be segmented as �you enjoy�)
+Input: �youleavefootprints�,
+Output: True (The string can be segmented as �you leave footprints� or �you leave foot prints�)
+Input:salmonenjoyapples
+Output: False
+
+//TEBOW IT
+/*
+1. T - TALK
+INPUT: Dictionary of String and a String
+OUTPUT: Boolean - True/False
+Clarified:
+Null input string or null dictionary - Exception
+Case sensitive -  Yes
+Special characters - No
+Empty string or Empty dictionary - Return False
+
+2. E - EXAMPLES
+2.1 Valid Inputs and corresponding expected outputs:
+
+	 	a. Dictionary { 'he', 'heart', 'cat' }, String = 'heartcat' -> TRUE
+	 	b. Dictionary { 'he', 'heart', 'art', 'artist' }, String = 'heartist' -> TRUE
+	 	c. Dictionary {' '}, String = ' ' -> FALSE
+
+2.2 Invalid Input and corresponding expected outputs:
+
+	  	a. Dictionary is NULL, String is NULL -> Exception
+
+3. B - BRUTE-FORCE
+
+3.1
+Find all possible ways of breaking up the input string. For each possible set of word breaks, check if one or more of them matches words in the dictionary.
+If yes, return true, else return false.
+Time complexity: O(2^n * m) where:
+n = length of input string
+m = length of word dictionary
+
+3.2
+Iterate over the entire length of the string character by character. If no match is found with the words in the dictionary even when you reach the end of the string, return false.
+If you found a match, repeat this same algorithm for the remaining portion of the input string (i.e. recursive call)
+
+Complexity
+Time: O(2^n)
+Space: O(n^2)
+
+4. O - OPTIMISE
+As we can see, the Word Break problem has an optimal substructure. Also, it can be broken down into multiple subproblems.
+Hence, we can use Dynamic Programming to solve this problem wherein the subproblems will be memoised rather than being computed again and again.
+-> Use an array for storing the results of the subproblems.
+-> -1: not encountered, 0: encountered, 1: can be segmented i.e, one or more of the word breaks of the string matches words in the dictionary.
+-> Consider all prefixes of the current string, If prefix is found in dictionary, then recurse for suffix.
+5. W - WALK THROUGH
+Explained as comments in the code
+6. I - IMPLEMENT
+*/
+#include <iostream>
+#include <string>
+#include <vector>
+#include <algorithm>
+using namespace std;
+bool wordBreak(vector<string> const &dict, string str, vector<int> &canBeSegmented)
```

Comment:
Another note: dict should be a type that has constant lookup time.  Especially if dict is very large, this could save some major time.  I think in c++ the best option is unordered_set, but there may be other options
https://stackoverflow.com/questions/921384/java-string-array-is-there-a-size-of-method

Reply:
Thanks for the link, I'll see if I could implement this.